velox/functions/prestosql/DateTimeFunctions.h

```diff
@@ -131,6 +131,37 @@ struct TimestampWithTimezoneSupport {
 
 } // namespace
 
+template <typename T>
+struct WeekFunction : public InitSessionTimezone<T>,
+                      public TimestampWithTimezoneSupport<T> {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE int64_t getWeek(const std::tm& time) {
+    int daysPerWeek = 7;
+    int64_t week = (time.tm_yday + daysPerWeek -
+                    (time.tm_wday ? (time.tm_wday - 1) : (daysPerWeek - 1))) /
```

Comment:
`daysPerWeek -
                    (time.tm_wday ? (time.tm_wday - 1) : (daysPerWeek - 1)` is hard to follow.
https://cplusplus.com/reference/ctime/tm/ says `tm_wday` is days since Sunday. Doesn't the above equation decrease as the value of `tm_wday` increases?

Reply:
Interesting point. Have you looked at the Presto code to calculate the same ?

https://en.wikipedia.org/wiki/ISO_week_date also has algorithms for calculating week of the year.

Reply:
Yes actually it is difficult to follow, I had a separate case for leap years earlier but then came across this implementation used in FreeBSD here: https://cplusplus.com/forum/general/220155/ . This seemed simpler and it worked correctly for edge cases so I followed the same. From what I understood, the above equation ensures the week with first Thursday is counted as week 1 by adding ``daysPerWeek - (time.tm_wday - 1)``. I will follow the algorithm on Wikipedia shared by Aditi since it is more clear, thanks!

Reply:
> 

Thanks for sharing, will follow the same

Reply:
In any case, what is the Presto java calculation ? We should be consistent with that in general.